### PR TITLE
Enhancement: Media and Documents

### LINKED ISSUE #962: Enhancement - Media/Documents
### Suggestion Lists

- [x] https://buddyboss.slack.com/archives/C013JBN4GRJ/p1591194344021800
- [x] https://buddyboss.slack.com/archives/C013JBN4GRJ/p1591343566000400
- [x] https://buddyboss.slack.com/archives/C013JBN4GRJ/p1591095578017100
- [ ] When you upload only 1 media/document and if you wriite any text then it should be added to that photo/document description. - Reverted the code as it creating the issue - ❌ 
- [x] https://buddyboss.slack.com/archives/DCCR02FDY/p1592966485000300 - https://www.loom.com/share/fd3aab2df4014b93991a06a54c412e90
- [x] Remove the activity from activity parent if it's move to other document folder
- [x] https://github.com/buddyboss/buddyboss-platform/issues/51
- [x] https://github.com/buddyboss/buddyboss-platform/issues/344
- [x] https://github.com/buddyboss/buddyboss-platform/issues/1332

### Proof of all the Changes 
https://github.com/buddyboss/buddyboss-platform/pull/1129#issuecomment-724595950
https://github.com/buddyboss/buddyboss-platform/pull/1129#issuecomment-724598829

### PR DESCRIPTION
**Pull request by chetansatasiya**
### All Submissions:

* [x] Have you followed the [Pull Requests Guidelines](https://github.com/buddyboss/buddyboss-platform/wiki/Submitting-Pull-Requests#pull-request-guidelines)?
* [x] Does your code follow the [WordPress' Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change?

### Changes proposed in this Pull Request:
Fixes #962

Connected PR in Theme:
https://github.com/buddyboss/buddyboss-theme/pull/893

### Other information:
* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

---
This change is [Reviewable](https://reviewable.io/reviews/buddyboss/buddyboss-platform/1129)

### CODE REVIEW

File: src/bp-activity/bp-activity-filters.php [hunk 1]
```diff
@@ -1737,6 +1737,17 @@ function bp_activity_create_parent_media_activity( $media_ids ) {
 			$activity_id = bp_activity_post_update( array( 'content' => $content ) );
 		}
 
+		// Add in description in attachment when only one document uploaded.
+		if ( ! empty( $activity_id ) && is_array( $media_ids ) && 1 === count( $media_ids ) && $content ) {
```
Comment (mehulkaklotar) [inline src/bp-activity/bp-activity-filters.php:1741]: @chetansatasiya can we check if `$content` is empty or not? right now it is only checking if `$content` variable is set.

File: src/bp-activity/bp-activity-filters.php [hunk 2]
```diff
@@ -1939,6 +1950,17 @@ function bp_activity_create_parent_document_activity( $document_ids ) {
 			$activity_id = bp_activity_post_update( array( 'content' => $content ) );
 		}
 
+		// Add in description in attachment when only one document uploaded.
+		if ( ! empty( $activity_id ) && is_array( $document_ids ) && 1 === count( $document_ids ) && $content ) {
```
Comment (mehulkaklotar) [inline src/bp-activity/bp-activity-filters.php:1954]: @chetansatasiya can we check if `$content` is empty or not? right now it is only checking if `$content` variable is set.

File: src/bp-document/bp-document-filters.php [hunk 3]
```diff
@@ -261,6 +261,20 @@ function bp_document_update_activity_document_meta( $content, $user_id, $activit
 		return false;
 	}
 
+	// Add in description in attachment when only one document uploaded.
+	if ( is_array( $_POST['document'] ) && 1 === count( $_POST['document'] ) ) {
+		if ( '' === trim( $content ) && ! empty( $activity_id ) ) {
+			$get_content = new BP_Activity_Activity( $activity_id );
+			$content = $get_content->content;
+		}
```
Comment (mehulkaklotar) [inline src/bp-document/bp-document-filters.php:269]: @chetansatasiya we don't need to get content from the activity here. we already get the content in function arguments `( $content, $user_id, $activity_id )` if in there `$content` is empty we dont go into the condition and add attachment content.

File: src/bp-media/bp-media-filters.php [hunk 4]
```diff
@@ -296,6 +296,20 @@ function bp_media_update_activity_media_meta( $content, $user_id, $activity_id )
 		return false;
 	}
 
+	// Add in description in attachment when only one media uploaded.
+	if ( is_array( $_POST['media'] ) && 1 === count( $_POST['media'] ) ) {
+		if ( '' === trim( $content ) && ! empty( $activity_id ) ) {
+			$get_content = new BP_Activity_Activity( $activity_id );
+			$content = $get_content->content;
+		}
```
Comment (mehulkaklotar) [inline src/bp-media/bp-media-filters.php:304]: @chetansatasiya we don't need to get content from the activity here. we already get the content in function arguments `( $content, $user_id, $activity_id )` if in there `$content` is empty we dont go into the condition and add attachment content.

Comment (chetansatasiya) [reply src/bp-media/bp-media-filters.php]: @mehulkaklotar DONE 👍

File: src/bp-document/bp-document-filters.php (hunk 3 above)
Comment (chetansatasiya) [reply src/bp-document/bp-document-filters.php]: @mehulkaklotar DONE 👍

File: src/bp-activity/bp-activity-filters.php (hunk 2 above)
Comment (chetansatasiya) [reply src/bp-activity/bp-activity-filters.php]: @mehulkaklotar DONE 👍

File: src/bp-activity/bp-activity-filters.php (hunk 1 above)
Comment (chetansatasiya) [reply src/bp-activity/bp-activity-filters.php]: @mehulkaklotar DONE 👍

**Comment by dipskakadiya** [comment]
@chetansatasiya @amincdev Can you resolve conflict?

**Comment by amincdev** [comment]
**Tasks included this PR:**

1. "Photos" Title missing in Group -> Photos tab while logged out
https://buddyboss.slack.com/archives/C013JBN4GRJ/p1591194344021800
https://www.loom.com/share/580707b106c848c5a5e3a101ef96212b

2. Create Folder/Album or choosing existing Option while uploading Photo / Document
https://buddyboss.slack.com/archives/C013JBN4GRJ/p1591343566000400

3. on "/photos" page "Upload photo" and "Create Album" buttons are missing
https://buddyboss.slack.com/archives/C013JBN4GRJ/p1591095578017100

4. When you upload only 1 media/document and if you wriite any text then it should be added to that photo/document description

5. Edit Forum cannot load added document or photo
https://buddyboss.slack.com/archives/DCCR02FDY/p1592966485000300
https://www.loom.com/share/fd3aab2df4014b93991a06a54c412e90\

6. Remove the activity from activity parent if it's move to other document folder

7. Ability to move photo outside of Album
https://github.com/buddyboss/buddyboss-platform/issues/51
Added move functionality

8. Option for activity photos to add them in any album
https://github.com/buddyboss/buddyboss-platform/issues/344
Added move functionality

9. Limit to add photos/document per batch
https://github.com/buddyboss/buddyboss-platform/issues/1332
Limit added already

10. Media Privacy same as document
 Prevent direct access via media URL same as Documents
 Prevent Download the media based on the privacy same as Documents

**Comment by amincdev** [comment]
**Task Explanation**
Part 1 - https://www.loom.com/share/90d9617e6209413bae0c4b128c8f3076
Part 2 - https://www.loom.com/share/95ae3627166a439f8a60cc183c1323cd
Part 3 - https://www.loom.com/share/c2e3c77f76404ec1bdda1640d1a814f4

**Comment by amincdev** [comment]
@mehulkaklotar Can you please review this?

**Comment by ankitguptaindia** [comment]
@amincdev @chetansatasiya @sourabhmatolia 


These newly added options are appearing out of image if image is small in size on Activity page.
https://www.loom.com/share/67723c28c741438597cd972a88e9e73c



---

When moving the media it shows Popup with list of all albums available. This popup shows `Medias` as title. this is not looking appropriate.
On the other hand In Documents popup, we don't have such type of titles.
https://www.loom.com/share/667a9d3cd481471a9dca642ba4f1a53f


----

Documents has validation check before deleting the files.

When user deleting the Documents, It ask for confirmation in alert box but this is not working in this way in Media.
https://www.loom.com/share/ccc7561387364e8da72f73e022ad4915

-----

More options of Media not taking theme typography settings.
https://www.loom.com/share/f19b0f9cad3b4afbac186b02b0527425


----

Can not Move & Delete Media if uploaded more than 4 in numbers.

Suppose, any user uploaded 10 photos and want to move last 3 photos to any album then can not do because we are showing more options only on grid not on single preview.

https://www.loom.com/share/8ec2c6889c51439ea40bbe1efa4d0274

 -------

Moving Media from activity to any folder creates new-new activity.
https://www.loom.com/share/b72e118fa58b48d3bd6491ab9617d77d


-----

Inconsistency between Documents Popup and MEdia popup.

Refer - https://www.loom.com/share/187d1a85e3ab4671b27f8d7996f1953e


----

All Photos grids become blank when moving photos from Grids view on activity page.
https://www.loom.com/share/fa21526d46b94cf68fb15bda734a0fe9

----

**Comment by amincdev** [comment]
> When moving the media it shows Popup with list of all albums available. This popup shows Medias as title. this is not looking appropriate.
On the other hand In Documents popup, we don't have such type of titles.

Fixed now: https://prnt.sc/vmsvkj

> Documents has validation check before deleting the files.
> When user deleting the Documents, It ask for confirmation in alert box but this is not working in this way in Media.

Fixed now: https://prnt.sc/vmswf8

> More options of Media not taking theme typography settings.

Fixed now (Loading from theme options): https://prnt.sc/vmsx85

> Can not Move & Delete Media if uploaded more than 4 in numbers.
> Suppose, any user uploaded 10 photos and want to move last 3 photos to any album then can not do because we are showing more options only on grid not on single preview.

This is intentional and already discuss with @eisenwasser before. User should be able to move those photos from the profile photo page.

> Inconsistency between Documents Popup and MEdia popup.

Fixed now: https://prnt.sc/vmqi26


> All Photos grids 
> become blank when moving photos from Grids view on activity page.

Fixed now


@ankitguptaindia Please take pull in platform and theme both. Also, re-save the theme options.
Working on remaining issues.

**Comment by ankitguptaindia** [comment]
@amincdev 


Search box is missing for photos and album page. 
Although we have the same functionality on the global photos page. Users can search photos from global photos page but not from Member Profile and Gruop photos page.

https://prnt.sc/vmn8cf 
https://prnt.sc/vmn968
https://i.imgur.com/FLA21E0.jpg

-----

If user reached max number of photos/files then it should give :

Proper detailed message like you can only upload max N number of file at a time instead of this incomplete message https://i.imgur.com/NRxLAWf.jpg
If user crossed the limit per allowed per batch then it should not give uploader to upload next file.
----

When creating a new Album from popup then it is not an option to select privacy.
I noticed it only happens when user select any already created album.
This way it sounds like the user is going to create a sub album in any already created album that's why it is not giving privacy options because it will take privacy from main album. 
But we don't have a Sub Album feature.
Please add conditions for this check. So it will always show Privacy option while creating new album from popup.


More details https://www.loom.com/share/e9cd97560d0b4842a4f98b8580e1dc84 

------

When user clicked delete icon without selecting any media then message need to style.
https://i.imgur.com/XOqNqkZ.jpg

---

If User selected multiple photo and deleted only single photo by using individual option then it is deleting all photos. It should only delete all when user click delete icon available on right top.
Please watch this video for better explanation https://www.loom.com/share/30cd4af1be724a17ac113e075dc5e536

-------
Suggestion -  If we add an option to delete albums from Album listing pages like we have added for delete photos and already have to delete documents then it would be easy to manage albums for users and will show consistency everywhere.

https://www.loom.com/share/980cc4350c9540e59a62627ff9b472e7


-----

UX -  If user already in any album and uploading new media then it should auto selected that album in popup.


https://www.loom.com/share/594a205fd67647d1acbaa77686d87801 


----

Suggestion -  please set min value to 1. Right not it is taking -ve values.

https://i.imgur.com/vOGhc5F.jpg

----

Minor- Typo add space  between File Size text https://i.imgur.com/1fTrWNd.jpg

---

cc @chetansatasiya @sourabhmatolia

**Comment by sourabhmatolia** [comment]
@amincdev 

- Please correct or redo style for move and delete dropdown here:
https://prnt.sc/vmwn42
Make it consistent. Ask for a design from Bilal.

- Go to Photos in Profile > Click on Delete Icon without selecting any image > See issue:
https://prnt.sc/vmwopn

- Go to Photos in Profile > Select image using check icon > Click on Delete.
It should show a standard "Are you sure" confirmation message before deleting.


cc: @ankitguptaindia

**Comment by chetansatasiya** [comment]
> Moving Media from activity to any folder creates new-new activity.
https://www.loom.com/share/b72e118fa58b48d3bd6491ab9617d77d 

@ankitguptaindia When you uploading a multiple media/document at a time then we are creating a 1 parent actvity and for all other uploaded media/document have it's own child activity and which is all together in parent actvity. So now once you move from that then we are making the child activity to as a parent and also we are giving the privacy based on the folder where you are moving OR we are giving the existing privacy when you moving into the root.

**Comment by amincdev** [comment]
@ankitguptaindia 

> Minor- Typo add space between File Size text
> https://i.imgur.com/1fTrWNd.jpg

**Fixed**

> When creating a new Album from popup then it is not an option to select privacy.
> I noticed it only happens when user select any already created album.
> This way it sounds like the user is going to create a sub album in any already created album that's why it is not giving privacy options because it will take privacy from main album.
> But we don't have a Sub Album feature.
> Please add conditions for this check. So it will always show Privacy option while creating new album from popup.
> More details
> https://www.loom.com/share/e9cd97560d0b4842a4f98b8580e1dc84

**Fixed**


> UX - If user already in any album and uploading new media then it should auto selected that album in popup.
> https://www.loom.com/share/594a205fd67647d1acbaa77686d87801

**Fixed.** User will get album selected:  https://prnt.sc/vo7esb


> When user clicked delete icon without selecting any media then message need to style.
> https://i.imgur.com/XOqNqkZ.jpg

**Fixed**

> Go to Photos in Profile > Select image using check icon > Click on Delete.
> It should show a standard "Are you sure" confirmation message before deleting.

**Fixed**

> Go to Photos in Profile > Click on Delete Icon without selecting any image > See issue:
> https://prnt.sc/vmwopn

**Fixed**

> If User selected multiple photo and deleted only single photo by using individual option then it is deleting all photos. It should only delete all when user click delete icon available on right top.
> Please watch this video for better explanation
> https://www.loom.com/share/30cd4af1be724a17ac113e075dc5e536

**Fixed**

> Suggestion - please set min value to 1. Right not it is takin

**Fixed**

> These newly added options are appearing out of image if image is small in size on Activity page.
> https://www.loom.com/share/67723c28c741438597cd972a88e9e73c
> Issue: https://prnt.sc/vo5h05

@sourabhmatolia For more than one photo we are already stretching image to fill the container. So it is fine. But for a single photo, We need some kind of design to make it look good without stretching full width. Like we did in document. Check screenshot: https://prnt.sc/vod2pf 
Let me know what you sugget.


**Working on Remaining issues.**

> If user reached max number of photos/files then it should give:
> Proper detailed message like you can only upload max N number of file at a time instead of this incomplete message https://i.imgur.com/NRxLAWf.jpg
> If user crossed the limit per allowed per batch then it should not give uploader to upload next file.

**Comment by chetansatasiya** [comment]
@ankitguptaindia @sourabhmatolia @amincdev @mohit-buddyboss 

**FYI**

Now I have removed the media URL changes from this branch and I made seprate branch for the media URL changes. Here is the branch for the media URL changes https://github.com/buddyboss/buddyboss-platform/pull/1779

So now in this PR media is still accessible publicy even if the media is private.

**Comment by amincdev** [comment]
@ankitguptaindia 

> If user reached max number of photos/files then it should give:
> Proper detailed message like you can only upload max N number of file at a time instead of this incomplete message https://i.imgur.com/NRxLAWf.jpg
> If user crossed the limit per allowed per batch then it should not give uploader to upload next file.

**Fixed**

**Proper message:**
https://prnt.sc/vp3f0o
https://prnt.sc/vp3fbl

**Does not show uploader after limit**
https://prnt.sc/vp3g09

**Comment by amincdev** [comment]
@ankitguptaindia 

> Search box is missing for photos and album page.
> Although we have the same functionality on the global photos page. Users can search photos from global photos page but not from Member Profile and Gruop photos page.
> 
> https://prnt.sc/vmn8cf
> https://prnt.sc/vmn968
> https://i.imgur.com/FLA21E0.jpg


As discussed, I have created a separate issue for this: https://github.com/buddyboss/buddyboss-platform/issues/1827

**Comment by ankitguptaindia** [comment]
@amincdev @chetansatasiya @sourabhmatolia 

FYI - 
As discussed the last scrum, I am adding new issues on respective Issue Link as this PR link has already many notification updates from Lint as comments. Maybe we miss any task due to huge notification.

https://github.com/buddyboss/buddyboss-platform/issues/962

**Comment by ankitguptaindia** [comment]
**STATUS of issues reported on this PR -** 


> These newly added options are appearing out of image if image is small in size on Activity page.
https://www.loom.com/share/67723c28c741438597cd972a88e9e73c

FIXED. 

---


>  When moving the media it shows Popup with list of all albums available. This popup shows `Medias` as title. this is not looking appropriate.
On the other hand In Documents popup, we don't have such type of titles.
https://www.loom.com/share/667a9d3cd481471a9dca642ba4f1a53f

FIXED> 

----


>  Documents has validation check before deleting the files.

When user deleting the Documents, It ask for confirmation in alert box but this is not working in this way in Media.
https://www.loom.com/share/ccc7561387364e8da72f73e022ad4915


FIXED>
-----


>  More options of Media not taking theme typography settings.
https://www.loom.com/share/f19b0f9cad3b4afbac186b02b0527425


FIXED.
----

 
>  Can not Move & Delete Media if uploaded more than 4 in numbers.
Suppose, any user uploaded 10 photos and want to move last 3 photos to any album then can not do because we are showing more options only on grid not on single preview.

https://www.loom.com/share/8ec2c6889c51439ea40bbe1efa4d0274


Okay then 

 -------


>  Moving Media from activity to any folder creates new-new activity.
https://www.loom.com/share/b72e118fa58b48d3bd6491ab9617d77d

Okay then 

-----


>  Inconsistency between Documents Popup and MEdia popup.

Refer - https://www.loom.com/share/187d1a85e3ab4671b27f8d7996f1953e

FIXED

----


>  All Photos grids become blank when moving photos from Grids view on activity page.
https://www.loom.com/share/fa21526d46b94cf68fb15bda734a0fe9

Fixed.

----




>  Search box is missing for photos and album page. 
Although we have the same functionality on the global photos page. Users can search photos from global photos page but not from Member Profile and Gruop photos page.

https://prnt.sc/vmn8cf 
https://prnt.sc/vmn968
https://i.imgur.com/FLA21E0.jpg


Okay.  a separate issue for this has been created: #1827

-----


>  If user reached max number of photos/files then it should give:

Proper detailed message like you can only upload max N number of file at a time instead of this incomplete message https://i.imgur.com/NRxLAWf.jpg
FIXED.




>  If user crossed the limit per allowed per batch then it should not give uploader to upload next file.

FIXED.


----


>  When creating a new Album from popup then it is not an option to select privacy.
I noticed it only happens when user select any already created album.
This way it sounds like the user is going to create a sub album in any already created album that's why it is not giving privacy options because it will take privacy from main album. 
But we don't have a Sub Album feature.
Please add conditions for this check. So it will always show Privacy option while creating new album from popup.
More details https://www.loom.com/share/e9cd97560d0b4842a4f98b8580e1dc84 

FIXED> 

------


>  When user clicked delete icon without selecting any media then message need to style.
https://i.imgur.com/XOqNqkZ.jpg

FIXED.


---


>  If User selected multiple photo and deleted only single photo by using individual option then it is deleting all photos. It should only delete all when user click delete icon available on right top.
Please watch this video for better explanation https://www.loom.com/share/30cd4af1be724a17ac113e075dc5e536

FIXED.

-------



>  Suggestion -  If we add an option to delete albums from Album listing pages like we have added for delete photos and already have to delete documents then it would be easy to manage albums for users and will show consistency everywhere.

https://www.loom.com/share/980cc4350c9540e59a62627ff9b472e7

MISSED.

@amincdev Seems you missed this. Please check. 

-----


>  UX -  If user already in any album and uploading new media then it should auto selected that album in popup.
  

https://www.loom.com/share/594a205fd67647d1acbaa77686d87801 

FIXED.


----


>  Suggestion -  please set min value to 1. Right not it is taking -ve values.

https://i.imgur.com/vOGhc5F.jpg

FIXED. 

----


>  Minor- Typo add space  between File Size text https://i.imgur.com/1fTrWNd.jpg

---

Fixed.

**Comment by amincdev** [comment]
>> Suggestion - If we add an option to delete albums from Album listing pages like we have added for delete photos and already have to delete documents then it would be easy to manage albums for users and will show consistency everywhere.
>> https://www.loom.com/share/980cc4350c9540e59a62627ff9b472e7

> MISSED.
> @amincdev Seems you missed this. Please check.

@ankitguptaindia I have created a separate issue for this: https://github.com/buddyboss/buddyboss-platform/issues/1871

File: src/bp-activity/bp-activity-template.php [hunk 5]
```diff
@@ -4184,26 +4160,29 @@ function bp_activity_entry_css_class() {
...
 	}
 
+	/**
+	 * Filters the determined classes to add to the HTML element.
+	 *
+	 * @since BuddyBoss 1.5.0
+	 *
+	 * @param string $value Classes to be added to the HTML element.
+	 */
+	return apply_filters( 'bp_get_activity_entry_css_class', $class );
```
Comment (mehulkaklotar) [inline src/bp-activity/bp-activity-template.php:4187]: @chetansatasiya use this filter in media and document component codebase and add classes there.

Comment (chetansatasiya) [reply src/bp-activity/bp-activity-template.php]: 👍

**Review by mehulkaklotar — CHANGES_REQUESTED** [review]
@chetansatasiya please go through all the changes i have shared with you and let me know i will approve here.

**Review by mehulkaklotar — APPROVED** [review]
@chetansatasiya @amincdev approved! 👍

**Comment by ankitguptaindia** [comment]
@amincdev @chetansatasiya @sourabhmatolia 

Issues reported on this PR and issues reported on this issue comment thread https://github.com/buddyboss/buddyboss-platform/issues/962 have been fixed. 

Checked all possible scenario and this PR is ready to merge 🚀  

Refer to connected issue link for more details and fixes status -  https://github.com/buddyboss/buddyboss-platform/issues/962

**Review by sourabhmatolia — APPROVED** [review]

**Pull request merged**